src/cudamatrix/cu-kernels.cu

```diff
@@ -251,6 +251,48 @@ static void _select_rows(const int* out_row_ptr, int* out_col_idx,
   }
 }
 
+// mat += alpha * smat
+//
+// We use warpSize threads per row to access only the nonzero elements.
+// Every CU1DBLOCK/warpSize rows share one thread block.
+// 1D grid to cover all rows of smat.
+template<typename Real>
+__global__
+static void _add_smat(Real* mat, MatrixDim mat_dim, Real alpha,
+                      const int* smat_row_ptr, const int* smat_col_idx,
+                      const Real* smat_val) {
+  const int i = blockIdx.x * blockDim.y + threadIdx.y; // row idx
+  if (i < mat_dim.rows) {
```

Comment:
I think the speed is bounded by un-coalesced memory access to the dense matrix, so it might not change a lot with CSR format.  Additional conversion with csr2coo() also takes time, which can be as long as _add_smat() itself, so may not be a good idea.  The nnz in a sparse matrix is usually much less compared to the number of elements in a typical dense matrix, so the sparse-mat-related kernel functions seems not critical to the over-all speed.

I will add a speed test for TraceMatSmat(), to compare with the old implementation. It has a similar access pattern with AddSmat().